src/fundus/scraping/html.py

```diff
@@ -274,99 +37,188 @@ class HTML:
-            )
+            else:
+                if filter_url(str(response.url)):
+                    basic_logger.debug(f"Skipped responded URL '{str(response.url)}' because of URL filter")
+                    continue
+                html = response.text
+
+                if response.history:
+                    basic_logger.info(f"Got redirected {len(response.history)} time(s) from {url} -> {response.url}")
+
+                source = (
+                    WebSourceInfo(self.publisher, type(self.url_source).__name__, self.url_source.url)
+                    if isinstance(self.url_source, URLSource)
+                    else SourceInfo(self.publisher)
+                )
+
+                yield HTML(
+                    requested_url=url,
+                    responded_url=str(response.url),
+                    content=html,
+                    crawl_date=datetime.now(),
+                    source=source,
+                )
+
+            if self.delay:
+                time.sleep(max(0.0, self.delay() - time.time() + timestamp))
+                timestamp = time.time()
+
+
+class CCNewsSource:
+    def __init__(self, *publishers: PublisherEnum, warc_path: str, headers: Optional[Dict[str, str]] = None):
+        self.publishers = publishers
+        self.warc_path = warc_path
+        self.headers = headers or _default_header
+
+        self._publisher_mapping: Dict[str, PublisherEnum] = {
+            urlparse(publisher.domain).netloc: publisher for publisher in publishers
+        }
+
+    def fetch(self, url_filter: Optional[URLFilter] = None) -> Iterator[HTML]:
+        def extract_content(record: WarcRecord) -> Optional[str]:
+            warc_body: bytes = record.reader.read()
+
+            try:
+                return str(warc_body, encoding=record.http_charset)
+            except (UnicodeDecodeError, TypeError):
+                encoding: Optional[str] = chardet.detect(warc_body)["encoding"]
+
+                if encoding is not None:
+                    basic_logger.debug(
+                        f"Trying to decode record {record.record_id!r} from {target_url!r} "
+                        f"using detected encoding {encoding}."
+                    )
+
+                    try:
+                        return str(warc_body, encoding=encoding)
+                    except UnicodeDecodeError:
+                        basic_logger.warning(
+                            f"Couldn't decode record {record.record_id!r} from {target_url!r} with "
+                            f"original charset {record.http_charset!r} using detected charset {encoding!r}."
+                        )
+                else:
+                    basic_logger.warning(
+                        f"Couldn't detect charset for record {record.record_id!r} from {target_url!r} "
+                        f"with invalid original charset {record.http_charset!r}."
+                    )
+
+            return None
+
+        with requests.Session() as session:
```

Comment:
Would it be appropriate to use the handled global session here, or is this a different use case?

Reply:
IMO since we are utilizing `multiprocessing` here, using the handled global session could be more confusing, suggesting that all processes would use the same object, when instead, they would all copy a process local version of the handled global session and operate on different objects.

Reply:
Yes, that makes sense. Thanks for the clarification.